common/auth/service_principal_configuration_provider.go

```diff
@@ -0,0 +1,47 @@
+package auth
+
+import (
+	"crypto/rsa"
+	"fmt"
+
+	"github.com/oracle/oci-go-sdk/common"
+)
+
+type servicePrincipalConfigurationProvider struct {
+	keyProvider *servicePrincipalKeyProvider
+	region      string
+}
+
+// NewServicePrincipalConfigurationProvider will create a new service principal configuration provider
+func NewServicePrincipalConfigurationProvider(tenancyID, region, cert, key string, intermediates []string, passphrase *string) (common.ConfigurationProvider, error) {
+	var err error
+	var keyProvider *servicePrincipalKeyProvider
+	if keyProvider, err = newServicePrincipalKeyProvider(tenancyID, region, cert, key, intermediates, passphrase); err != nil {
+		return nil, fmt.Errorf("failed to create a new key provider: %s", err.Error())
+	}
+	return servicePrincipalConfigurationProvider{keyProvider: keyProvider, region: region}, nil
+}
+
+func (p servicePrincipalConfigurationProvider) PrivateRSAKey() (*rsa.PrivateKey, error) {
+	return p.keyProvider.PrivateRSAKey()
+}
+
+func (p servicePrincipalConfigurationProvider) KeyID() (string, error) {
+	return p.keyProvider.KeyID()
+}
+
+func (p servicePrincipalConfigurationProvider) TenancyOCID() (string, error) {
+	return "", nil
```

Comment:
I was just talking with @kohashim about this, I think it is better to return an error here, given that this provider does not actually support tenancy

Reply:
This allows to use the ComposingConfigProvider like this:
ComposingProvider(ServicePrincipasConfigProvider{}, FileConfigProvider{}) presumably the last config provider would provide the rest of the settings that ServicePrincipals does not

Reply:
This makes sense but will cause validation to fail when instantiating a client from the configuration provider.

IsConfigurationProviderValid Tests all parts of the configuration provider do not return an error

Reply:
Right, you know what let's leave it as is for now, I'll refactor that change in the future